tun/tun_illumos.go

```diff
@@ -0,0 +1,355 @@
+//go:build illumos && amd64
+// +build illumos,amd64
+
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright (C) 2017-2021 WireGuard LLC. All Rights Reserved.
+ * Copyright 2019 Joyent, Inc.
+ */
+
+package tun
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	tunNode = "/dev/tun"
+)
+
+type NativeTun struct {
+	fd      int
+	name    string
+	tunFile *os.File
+	events  chan Event
+	errors  chan error
+}
+
+// XXX This is largely copied from the freebsd code...
+// XXX How do I test if it is actually working?
+func (tun *NativeTun) routineRouteListener() {
+	var (
+		statusUp  bool
+		statusMTU int
+	)
+
+	tunIfindex, err := func() (int, error) {
+		iface, err := net.InterfaceByName(tun.name)
+		if err != nil {
+			return -1, err
+		}
+		return iface.Index, nil
+	}()
+	if err != nil {
+		tun.errors <- err
+		return
+	}
+
+	defer close(tun.events)
```

Comment:
Since the tunIfIndex logic can error out, you want to move this defer above it, so the channel closes regardless of which error path causes an exit.